interchain-token-service/src/contract/execute.rs

```diff
@@ -0,0 +1,125 @@
+
+    match HubMessage::abi_decode(&payload).change_context(Error::InvalidPayload)? {
+        HubMessage::SendToHub {
+            destination_chain,
+            message,
+        } => {
+            let destination_address = load_its_contract(deps.storage, &destination_chain)
+                .change_context_lazy(|| Error::UnknownChain(destination_chain.clone()))?;
+
+            let destination_payload = HubMessage::ReceiveFromHub {
+                source_chain: cc_id.source_chain.clone(),
+                message: message.clone(),
+            }
+            .abi_encode();
+
+            Ok(send_to_destination(
+                deps.storage,
+                deps.querier,
+                destination_chain.clone(),
+                destination_address,
+                destination_payload,
+            )?
+            .add_event(
+                Event::MessageReceived {
+                    cc_id,
+                    destination_chain,
+                    message,
+                }
+                .into(),
+            ))
+        }
+        _ => bail!(Error::InvalidMessageType),
+    }
+}
+
+fn normalize(chain: &ChainNameRaw) -> ChainName {
+    ChainName::try_from(chain.as_ref()).expect("invalid chain name")
+}
+
+/// Ensures that the source address of the cross-chain message is the registered ITS contract for the source chain.
+fn ensure_its_source_address(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    source_address: &Address,
+) -> Result<(), Error> {
+    let source_its_contract = load_its_contract(storage, source_chain)
+        .change_context_lazy(|| Error::UnknownChain(source_chain.clone()))?;
+
+    ensure!(
+        source_address == &source_its_contract,
+        Error::UnknownItsContract(source_address.clone())
+    );
+
+    Ok(())
+}
+
+fn send_to_destination(
+    storage: &dyn Storage,
+    querier: QuerierWrapper,
+    destination_chain: ChainNameRaw,
+    destination_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    let config = load_config(storage);
+
+    let gateway: axelarnet_gateway::Client =
+        client::Client::new(querier, &config.axelarnet_gateway).into();
+
+    let call_contract_msg =
+        gateway.call_contract(normalize(&destination_chain), destination_address, payload);
+
+    Ok(Response::new().add_message(call_contract_msg))
+}
+
+pub fn register_its_contract(
+    deps: DepsMut,
+    chain: ChainNameRaw,
+    address: Address,
+) -> Result<Response, Error> {
+    state::save_its_contract(deps.storage, &chain, &address)
+        .change_context_lazy(|| Error::FailedItsAddressRegistration(chain.clone()))?;
+
+    Ok(Response::new().add_event(Event::ItsContractRegistered { chain, address }.into()))
+}
+
+pub fn deregister_its_contract(deps: DepsMut, chain: ChainNameRaw) -> Result<Response, Error> {
+    state::remove_its_contract(deps.storage, &chain);
+
+    Ok(Response::new().add_event(Event::ItsContractDeregistered { chain }.into()))
```

Comment:
This doesn't seem to fail if the chain isn't registered in the first place. Wouldn't the event be confusing then?

Reply:
Good catch, added a check since I'd like to keep the event